include/ccf/kv/map_diff.h

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/kv/untyped_map_diff.h"
+
+namespace kv
+{
+  template <typename K, typename V, typename KSerialiser, typename VSerialiser>
+  class MapDiff : public AbstractHandle
+  {
+  protected:
+    kv::untyped::MapDiff map_diff;
+
+  public:
+    using KeyType = K;
+    using ValueType = V;
+
+    MapDiff(kv::untyped::MapDiff map_diff_) : map_diff(map_diff_) {}
+
+    MapDiff(kv::untyped::ChangeSet& changes, const std::string& map_name) :
+      map_diff(changes, map_name)
+    {}
+
+    /** Get value for key.
+     *
+     * @param key Key to read
+     *
+     * @return nullopt if key does not exist, optional<nullopt> if key exists
+     * but was deleted and the value if it exists and was not deleted.
+     */
+    std::optional<std::optional<V>> get(const K& key)
+    {
+      const auto opt_v_rep = map_diff.get(KSerialiser::to_serialised(key));
+
+      if (opt_v_rep.has_value())
+      {
+        if (opt_v_rep.value().has_value())
+        {
+          return VSerialiser::from_serialised(opt_v_rep.value().value());
+        }
+      }
+
+      return std::nullopt;
+    }
+
+    /** Test if key is present and not deleted.
+     *
+     * @param key Key to read
+     *
+     * @return Boolean true iff key exists
+     */
+    bool has(const K& key)
+    {
+      return map_diff.has(KSerialiser::to_serialised(key));
+    }
+
+    /** Test if key is deleted in this diff.
+     *
+     * @param key Key to read
+     *
+     * @return Boolean true iff key was deleted
+     */
+    bool is_deleted(const K& key)
+    {
+      return map_diff.is_deleted(KSerialiser::to_serialised(key));
+    }
+
+    /** Iterate over all entries in the map.
+     *
+     * The passed functor should have the signature
+     * `bool(const K& k, const std::optional<V>& v)`.
```

Comment:
Perhaps worth mentioning that a nullopt indicates a key being deleted in that diff?